shell/testing/tester_main.cc

```diff
@@ -513,6 +524,94 @@ int RunTester(const flutter::Settings& settings,
+}
+
+EXPORTED Dart_Handle LookupEntryPoint(const char* uri, const char* name) {
+  if (!uri || !name) {
+    return Dart_Null();
+  }
+  return Dart_GetField(Dart_LookupLibrary(Dart_NewStringFromCString(uri)),
+                       Dart_NewStringFromCString(name));
+}
+
+EXPORTED void Spawn(const char* entrypoint, const char* route) {
+  auto shell = g_shell->get();
+  fml::AutoResetWaitableEvent latch;
+  auto isolate = Dart_CurrentIsolate();
+  Dart_ExitIsolate();
+  auto spawn_task = [shell, &latch, entrypoint = std::string(entrypoint),
+                     route = std::string(route)]() {
+    auto configuration = RunConfiguration::InferFromSettings(
+        shell->GetSettings(), /*io_worker=*/nullptr,
+        /*launch_type=*/IsolateLaunchType::kExistingGroup);
+    configuration.SetEntrypoint(entrypoint);
+
+    Shell::CreateCallback<PlatformView> on_create_platform_view =
+        fml::MakeCopyable([](Shell& shell) mutable {
+          ImpellerVulkanContextHolder impeller_context_holder;
+          return std::make_unique<TesterPlatformView>(
+              shell, shell.GetTaskRunners(),
+              std::move(impeller_context_holder));
+        });
+
+    Shell::CreateCallback<Rasterizer> on_create_rasterizer = [](Shell& shell) {
+      return std::make_unique<Rasterizer>(
+          shell, Rasterizer::MakeGpuImageBehavior::kBitmap);
+    };
+
+    // Spawn a shell, and keep it running until it has no live ports, then
+    // delete it on the platform thread.
+    auto spawned_shell =
+        shell
+            ->Spawn(std::move(configuration), route, on_create_platform_view,
+                    on_create_rasterizer)
+            .release();
+
+    ConfigureShell(spawned_shell);
+
+    fml::TaskRunner::RunNowOrPostTask(
+        spawned_shell->GetTaskRunners().GetUITaskRunner(), [spawned_shell]() {
+          fml::MessageLoop::GetCurrent().AddTaskObserver(
+              reinterpret_cast<intptr_t>(spawned_shell), [spawned_shell]() {
+                if (spawned_shell->EngineHasLivePorts()) {
+                  return;
+                }
+
+                fml::MessageLoop::GetCurrent().RemoveTaskObserver(
+                    reinterpret_cast<intptr_t>(spawned_shell));
+                // Shell must be deleted on the platform task runner.
+                fml::TaskRunner::RunNowOrPostTask(
+                    spawned_shell->GetTaskRunners().GetPlatformTaskRunner(),
+                    [spawned_shell]() { delete spawned_shell; });
+              });
+        });
+    latch.Signal();
+  };
+  // The global shell pointer is kept alive by the RunTester routine above.
+  // This in turn keeps the isolate group alive. The UI task runner is blocked
+  // until the isolate spawns, which guarantees the global shell pointer won't
+  // be deleted.
+  fml::TaskRunner::RunNowOrPostTask(
+      shell->GetTaskRunners().GetPlatformTaskRunner(), spawn_task);
+  latch.Wait();
+  Dart_EnterIsolate(isolate);
```

Comment:
It seems like the reason you have to exit and re-enter the isolate is because the callback gets executed synchronously. Do you then really need to use `RunNowOrPostTask()` - can you just call the lambda here yourself?

Reply:
Yes, the process can be invoked in such a way that it uses multiple threads - if it does, this makes sure the call is executed on the correct thread.